### PR TITLE
[Refactor] `SetNameValuePair`

### PR DESCRIPTION
**Pull request by danieldoglas**
### Details

Changes to NVP calls on App to use specific methods.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/211821

### Tests

- [x] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [X] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [x] MacOS / Chrome
    - [ ] MacOS / Desktop
- [x] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose and it is
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn’t already exist
    - [x] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

<details>
<summary><h4>PR Reviewer Checklist</h4></summary>

- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there’s a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained “why” the code was doing something instead of only explaining “what” the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named “index.js”. All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/PR_REVIEW_GUIDELINES.md)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn’t already exist
    - [ ] The style can’t be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.

</details>

### QA Steps

Open your network tab in developer tools before running the tests

-  SetPreferredEmojiSkinTone
1. Access any chat you prefer
2. Click on Emoji
3. Click on `Change default skin tone`
4. Select a tone that is different that the current one
5. Check if a request was made to `/api?command=UpdatePreferredEmojiSkinTone`
6. Sign off and then Sign in again
7. Verify that skin tone is the same previously selected

-  SetFrequentlyUsedEmojis
1. Access any chat you prefer
2. Click on Emoji
3. Verify the current set of frequently used emoji and click on one that is not on that list
4. Check if a request was made to `/api?command=UpdateFrequentlyUsedEmojis`
5. Verify that your current list of frequently used was updated with the emoji you just used
6. Sign off and then Sign in again
7. Check if list is still the same.

-  SetChatPriorityMode
1. Access Settings > Preferences
2. Change “Priority Mode”
3. Verify that a request was made to `/api?command=UpdateChatPriorityMode`
4. Verify that your chat list exhibition mode was changed
5. Sign off and then Sign in again
6. Access Settings > Preferences
7. Verify that Priority Mode was the one you selected before

-  AddPaypalMeAddress
1. Access Settings > Payments
2. Click on “Add Payment method”
3. Select [Paypal.me](http://paypal.me/)
4. Add a handle e.g. “danielsilva”
5. Verify that a request was made to `/api?command=AddPaypalMeAddress`
6. Verify that the PayPal address is now listed as a payment method
7. Sign off and then Sign in again
8. Access Settings > Payments
9. Verify that PayPal is still listed

-  DeletePaypalMeAddress
1. Access Settings > Payments
2. Click on your PayPal option and then “Delete” and “Delete” again
3. Verify that a request was made to `/api?command=DeletePaypalMeAddress`
4. Verify that the PayPal address is not listed as a payment method
5. Sign off and then Sign in again
6. Access Settings > Payments
7. Verify that PayPal is not listed

- [ ] Verify that no errors appear in the JS console

#### Web

https://user-images.githubusercontent.com/5818999/178586419-5e223ab3-3016-4c6d-b1e2-9d9dacc93f6c.mov

### CODE REVIEW

File: src/libs/actions/User.js [hunk 1]
```diff
@@ -268,6 +267,45 @@ function isBlockedFromConcierge(blockedFromConcierge) {
...
+ */
+function deletePaypalMeAddress() {
+    const optimisticData = [
+        {
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: ONYXKEYS.NVP_PAYPAL_ME_ADDRESS,
+            value: '',
+        },
+    ];
+    API.write('DeletePaypalMeAddress', {}, {optimisticData});
```
Comment (sketchydroide) [inline src/libs/actions/User.js:299]: even if there is no value it feels weird to have this in a single line, unless linter asks it to be like this I would have in multiple lines

**Review by sketchydroide — DISMISSED** [review]
add a couple of questions

**Comment by sketchydroide** [comment]
It looks good to me, apart from a few questions, but I do feel to confident in my knowledge of the API refactor without asking for someone with more knowledge to give it a review as well.
Not sure if @puneetlath  is ok with being that person.

File: src/libs/actions/User.js [hunk 2]
```diff
@@ -362,15 +400,60 @@ function subscribeToExpensifyCardUpdates() {
...
+        value: JSON.stringify(frequentlyUsedEmojis),
+    }, {optimisticData});
+}
+
+/**
+ * Sync preferred timezone with Onyx and Server
+ * @param {String} timezone
+ */
+function setPreferredTimezone(timezone) {
+    API.write('SetTimezone', {
```
[the remark this answers is not in the document]
Comment (danieldoglas) [reply src/libs/actions/User.js:437]: Nope. The timezone data is stored with the Profile information, so it's being saved somewhere else!

File: src/libs/actions/PersonalDetails.js [hunk 3]
```diff
@@ -278,7 +278,7 @@ function setPersonalDetails(details, shouldGrowl) {
         .then((response) => {
             if (response.jsonCode === 200) {
                 if (details.timezone) {
-                    NameValuePair.set(CONST.NVP.TIMEZONE, details.timezone);
+                    User.setPreferredTimezone(details.timezone);
```
Comment (marcaaron) [inline src/libs/actions/PersonalDetails.js:281]: This seems like "chaining" and not something we want. Is `PersonalDetails_Update` being handled elsewhere? I would expect to see:

- `API.write('UpdatePersonalDetails')`
- And on the server the timezone would be set and not done in a separate API call

File: src/libs/actions/User.js [hunk 4]
```diff
@@ -268,6 +267,45 @@ function isBlockedFromConcierge(blockedFromConcierge) {
...
+    API.write('DeletePaypalMeAddress', {}, {optimisticData});
+    Growl.show(Localize.translateLocal('paymentsPage.deletePayPalSuccess'), CONST.GROWL.SUCCESS, 3000);
+}
+
+/**
+ * Fetch the public domain info for the current user.
+ *
+ * This API is a bit weird in that it sometimes depends on information being cached in bedrock.
+ * If the info for the domain is not in bedrock, then it creates an asynchronous bedrock job to gather domain info.
+ * If that happens, this function will automatically retry itself in 10 minutes.
```
Comment (marcaaron) [inline src/libs/actions/User.js:308]: This seems unrelated to the current PR.

File: src/libs/actions/User.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (marcaaron) [reply src/libs/actions/User.js:437]: Good sign that we shouldn't be doing this here at all and that it belongs in the server.

File: src/libs/actions/User.js [hunk 5]
```diff
@@ -362,15 +400,60 @@ function subscribeToExpensifyCardUpdates() {
...
+    API.write('SetTimezone', {
+        value: JSON.stringify(timezone),
+    }, {});
+}
+
+/**
+ * Sync user chat priority mode with Onyx and Server
+ * @param {String} mode
+ */
+function setChatPriorityMode(mode) {
```
Comment (marcaaron) [inline src/libs/actions/User.js:446]:
```suggestion
function updateChatPriorityMode(mode) {
```

File: src/libs/actions/User.js [hunk 6]
```diff
@@ -362,15 +400,60 @@ function subscribeToExpensifyCardUpdates() {
...
+    API.write('SetFrequentlyUsedEmojis', {
+        value: JSON.stringify(frequentlyUsedEmojis),
+    }, {optimisticData});
+}
+
+/**
+ * Sync preferred timezone with Onyx and Server
+ * @param {String} timezone
+ */
+function setPreferredTimezone(timezone) {
```
Comment (marcaaron) [inline src/libs/actions/User.js:436]:
```suggestion
function updatePreferredTimezone(timezone) {
```

File: src/libs/actions/User.js [hunk 7]
```diff
@@ -362,15 +400,60 @@ function subscribeToExpensifyCardUpdates() {
  * @param {String} skinTone
  */
 function setPreferredSkinTone(skinTone) {
```
Comment (marcaaron) [inline src/libs/actions/User.js:402]:
```suggestion
function updatePreferredSkinTone(skinTone) {
```

**Review by marcaaron — CHANGES_REQUESTED** [review]
I could be wrong, but I thought we decided to standardize on:

- Add
- Delete
- Update

And would not be using `Set*` or `set*` naming style as it doesn't reflect what the user is doing...

https://expensify.slack.com/archives/C02HWMSMZEC/p1655382955578269?thread_ts=1655296070.103369&cid=C02HWMSMZEC

File: src/libs/actions/User.js [hunk 8]
```diff
@@ -362,15 +400,60 @@ function subscribeToExpensifyCardUpdates() {
...
+    API.write('SetPreferredEmojiSkinTone', {
+        value: skinTone,
+    }, {optimisticData});
 }
 
 /**
  * Sync frequentlyUsedEmojis with Onyx and Server
  * @param {Object[]} frequentlyUsedEmojis
  */
 function setFrequentlyUsedEmojis(frequentlyUsedEmojis) {
```
Comment (puneetlath) [inline src/libs/actions/User.js:419]:
```suggestion
function updateFrequentlyUsedEmojis(frequentlyUsedEmojis) {
```

Similar to Marc's suggestion below.

**Review by puneetlath — CHANGES_REQUESTED** [review]

**Comment by danieldoglas** [comment]
ok, considering the name structure (my bad), I think we need to do those changes in Web-E to set those to the right naming convention before changing just the JS methods here. @marcaaron should I do that and put this on HOLD again?

**Comment by marcaaron** [comment]
Yes, I think that would be best! Otherwise we will have an App version that is trying to use API commands that have been removed. Just easier to update in Web-E then do it here after.

File: src/libs/actions/User.js (hunk 4 above)
Comment (danieldoglas) [reply src/libs/actions/User.js]: removed, merge error

File: src/libs/actions/User.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (danieldoglas) [reply src/libs/actions/User.js]: Removed this and reverted the changes. Agree that should be addressed in the ProfileUpdate

File: src/libs/actions/PersonalDetails.js (hunk 3 above)
Comment (danieldoglas) [reply src/libs/actions/PersonalDetails.js]: Removed this change

**Comment by danieldoglas** [comment]
Changes were addressed, PR created for Web-E. This is now on HOLD until web-e is deployed.

**Review by puneetlath — APPROVED** [review]

File: src/libs/actions/User.js [hunk 9]
```diff
@@ -267,6 +266,40 @@ function isBlockedFromConcierge(blockedFromConcierge) {
...
+        {
+            onyxMethod: CONST.ONYX.METHOD.MERGE,
+            key: ONYXKEYS.NVP_PAYPAL_ME_ADDRESS,
+            value: address,
+        },
+    ];
+    API.write('AddPaypalMeAddress', {
+        value: address,
+    }, {optimisticData});
+}
```
Comment (marcaaron) [inline src/libs/actions/User.js:285]: I'm actually curious why we are double storing the paypalme address... ?

https://github.com/Expensify/App/blob/13d2766a1988d5119fda6945a16b1eca32594267/src/libs/actions/PersonalDetails.js#L100-L111

I think maybe we would just refer to `personalDetails[currentUserEmail].payPalMeAddress` everywhere. I don't have the full context on why we are storing this in two places.

Comment (marcaaron) [reply src/libs/actions/User.js:285]: Maybe we can combine them (also maybe makes sense to do this in a follow-up though - not sure)

**Review by marcaaron — APPROVED** [review]
LGTM! 

Random thought... as I was reviewing this, I was thinking that a lot of these "user level" NVPs could just get stored on the `personalDetails[currentUserLogin]` object. Just a thought though, not sure if it's worth doing.

File: src/libs/actions/User.js (hunk 9 above)
Comment (danieldoglas) [reply src/libs/actions/User.js:285]: I also agree that this should be a follow up. Considering we have Paypal as part of the payment methods, we shouldn't be calling "UpdatePersonalDetails" from that screen. we can unify it, but we need to decide where it's going to be the better place.
